### PR TITLE
Feature: Allow users to comment on files

### PR DESCRIPTION
**Pull request by samanehsan**
From PR #2534, addresses JIRA ticket OSF-4300

Changes
------------
<b>OSF Changes:</b>
- Allow users to comment on their project files (any provider)
- When a user adds, edits, deletes or restores a comment it will show up in the project logs. 
- The total number of unread file comments appears next to the files tab and the number of unread comments on the project overview page appears next to the project title in the nav bar. 
- Uses API v2 routes for comments

<b>API Changes</b>: 
- Update comments serializer to allow commenting on files
- Update node comments list view to return all comments for that node (this includes project comments, file comments and replies)
- Node serializer comment relationship field unread comment count includes 'total', 'node', and (total) 'files'
- Modify Relationshipfield to take in a filter query in the format `{field_name: <value>}` and return a link that has that filter.
	- The comment replies relationship field now uses the format `nodes/{node_id}/comments/?filter[target]=comment_id` and the comment replies endpoint was removed
- File serializer:
	- `FileCommentRelationshipField` will only return comment relationship field if kind is not 'folder'
	- The link is in the format `nodes/{node_id}/comments/?filter[target]=file_id`
	- The meta info includes the number of unread comments on that file
- Fix filtering on a serializer method field -- `convert_key` will change the field to `*` (the field source), but it should be the field name.

Post-Merge
---------------
Run `scripts/update_comments.py` to update format of `User.comments_viewed_timestamp`

### CODE REVIEW

**Comment by sloria** [comment]
There are still multiple requests being sent when re-expanding comments:

comment_requests2

**Comment by sloria** [comment]
The "Delete" tooltip does not word-break correctly:

osf___comments

Unless the tooltips were in the spec or @erinmayhood objects, I think we can get rid of the tooltips. The functions of the buttons are already clear.

**Comment by samanehsan** [comment]
@sloria, it looks like one is an `OPTIONS` request and the second is the `GET` request. Is there a way to prevent the `OPTIONS` request from getting sent?

**Comment by sloria** [comment]
The preflight OPTIONS request is necessary. However, once the user has requested replies for a comment, they shouldn't need to request them again.

**Comment by samanehsan** [comment]
Ok, I see what you mean

**Comment by sloria** [comment]
That said, the above change is subject to the "5-minute" rule. I see that the undesired behavior is also on production.

**Comment by sloria** [comment]
osf_____foo_txt

1. The "+" on the deleted comment is a different color from a non-deleted comment. They should both be black.
2. The cursor for the "+" on a deleted comment should be a pointer.
3. The position for "+" buttons should be the same for both deleted and non-deleted comments, i.e. both on the right side.

These are minor and can go in a follow-up PR.

File: website/addons/base/views.py [hunk 1]
```diff
@@ -568,6 +568,9 @@ def addon_view_file(auth, node, file_node, version):
         'size': version.size if version.size is not None else 9966699,
         'private': getattr(node.get_addon(file_node.provider), 'is_private', False),
         'file_tags': [tag._id for tag in file_node.tags],
+        'file_id': file_node._id,
+        # Note: Comments are allowed for all addons that support waterbutler
+        'allow_comment': True
```
Comment (sloria) [inline website/addons/base/views.py]: Will this ever be `False`?

Comment (samanehsan) [inline website/addons/base/views.py]: It used to be `False` for dataverse, but then it got switched over to use waterbutler. If all future storage addons will use waterbutler, then I think we can get rid of this.

Comment (sloria) [inline website/addons/base/views.py]: OK nice.

**Comment by sloria** [comment]
I'm finishing my pass here. No blockers above. I'll merge and deploy to staging later this eve or first thing in the morning. If you can address the comments above, great; otherwise, they can be in a follow-up PR.

**Comment by samanehsan** [comment]
I've addressed everything except for removing the tooltips, removing the `allow_comment` flag and UI fixes to the toggle replies button. @sloria, I can make these improvements in a follow-up PR or in this one. :tada:

**Comment by sloria** [comment]
Merged! :octopus:  Will deploy to staging in the morning.

**Pull request merged**